### PR TITLE
Simple data writer

### PR DESCRIPTION
**Pull request by akshayn1107**
This data writer simply writes bytes to a file. Useful if no data transformations are required. One thing to consider in this PR is if we should refactor the writers to reduce code duplication, since this is really the first non-avro based writer we are pushing up.

Code Review: @eogren @vikarmic

### CODE REVIEW

File: gobblin-core/src/main/java/gobblin/writer/SimpleDataWriter.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+    this.outputFile = new Path(outputDir, fileName);
+    // Create the parent directory of the output file if it does not exist
+    if (!this.fs.exists(this.outputFile.getParent())) {
+      this.fs.mkdirs(this.outputFile.getParent());
+    }
+    this.recordsWritten = 0;
+    this.bytesWritten = 0;
+    this.closed = false;
+  }
```
Comment (vikarmic) [inline gobblin-core/src/main/java/gobblin/writer/SimpleDataWriter.java]: If anything's a candidate for refactoring, it's probably this constructor.  Most of these properties are going to be common to most writers, right?

Comment (eogren) [inline gobblin-core/src/main/java/gobblin/writer/SimpleDataWriter.java]: I think I have a refactor of this sitting around in my fork somewhere - I can take a look.

**Comment by akshayn1107** [comment]
@eogren @vikarmic I've refactored some of the common code. Check it out and let me know what you think.

File: gobblin-core/src/main/java/gobblin/writer/SimpleDataWriter.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+  private int recordsWritten;
+  private int bytesWritten;
+  private boolean closed;
+
+  public SimpleDataWriter(State properties, String fileName, int numBranches, int branchId) throws IOException {
+    super(properties, fileName, numBranches, branchId);
+    String delim;
+    if ((delim = properties.getProp(ConfigurationKeys.SIMPLE_WRITER_DELIMITER, null)) == null || delim.length() == 0) {
+      recordDelimiter = null;
```
Comment (eogren) [inline gobblin-core/src/main/java/gobblin/writer/SimpleDataWriter.java]: just for consistency's sake, would probably say this.recordDelimiter = null (and same thing if it's not null)

File: gobblin-core/src/main/java/gobblin/writer/SimpleDataWriter.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    if (!this.fs.exists(this.stagingFile)) {
+      throw new IOException(String.format("File %s does not exist", this.stagingFile));
+    }
+
+    LOG.info(String.format("Moving data from %s to %s", this.stagingFile, this.outputFile));
+    // For the same reason as deleting the staging file if it already exists, deleting
+    // the output file if it already exists prevents task retry from being blocked.
+    if (this.fs.exists(this.outputFile)) {
+      LOG.warn(String.format("Task output file %s already exists", this.outputFile));
+      this.fs.delete(this.outputFile, false);
```
Comment (eogren) [inline gobblin-core/src/main/java/gobblin/writer/SimpleDataWriter.java]: Do you know if we need to check the return code here? Kind of unclear from the Hadoop javadocs https://hadoop.apache.org/docs/current/api/org/apache/hadoop/fs/FileSystem.html#delete(org.apache.hadoop.fs.Path, boolean) -- I read that as it throws IOException ONLY if recursive is set to true, but maybe their docs are just bad?

**Comment by eogren** [comment]
otherwise looks good to me!

File: gobblin-core/src/main/java/gobblin/writer/SimpleDataWriter.java (hunk 3 above)
Comment (akshayn1107) [inline gobblin-core/src/main/java/gobblin/writer/SimpleDataWriter.java]: I actually don't know, but just to be safe I wrote a wrapper around it which throws IOException if the delete function returns false.

Comment (eogren) [inline gobblin-core/src/main/java/gobblin/writer/SimpleDataWriter.java]: great, thanks!

**Pull request merged**